### DISCUSSION TITLE
Another darglint replacement

### DISCUSSION THREAD

**Discussion by jsh9** [jsh9/pydoclint#9] [General]
Hi @akaihola and @real-yfprojects:

I was a user of darglint.  I found some aspects of darglint a dealbreaker for me, so a few weeks ago I wrote a new linter, **pydoclint**, from scratch (https://github.com/jsh9/pydoclint).

It was not until today that I found out about darglint2 that you started (also recently).

That's why I'm starting this discussion thread to let you know of my linter.

In any case, I plan to keep maintaining and improving **pydoclint**, and I have no intention to contribute to darglint or any of its derivatives, for two main reasons:

1. darglint doesn't support `# noqa: ...` with flake8 well
2. darglint is too slow; here's a benchmarking test I did:

|                                                              | pydoclint | darglint                          |
| ------------------------------------------------------------ | --------- | --------------------------------- |
| [numpy](https://github.com/numpy/numpy)                      | 2.0 sec   | 49 min 9 sec (1,475x slower)      |
| [scikit-learn](https://github.com/scikit-learn/scikit-learn) | 2.4 sec   | 3 hr 5 min 33 sec (4,639x slower) |

**Comment by real-yfprojects** [comment]
Thanks for pinging me. Your tool looks nice.

> It was not until today that I found out about darglint2 that you started (also recently).

Yeah, we are working on being mentioned in the original darglint2 repository. We can definitely mention your project in our docs.

In contrast to pydoclint darglint2 isn't a new linter or a new tool but a fork of darglint. We basically took over maintaining the project since we found it valuable. 

I also noticed that performance is almost a deal breaking issue with darglint (or darglint2) which we will definitely have to face when we decide to keep the project going. The usefulness of darglint2 has to be reevaluated now that there is a faster alternative. Can you maybe provide a comparison of supported lint rules? For that matter I would also be interested in the other features you have planned and the philosophy/vision of your tool. 

For now I think there might be value in darglint2 because I have big plans on extending the linting capabilities to class and module docstrings.

 
> 	|pydoclint 	| darglint | |
> |----| ---|----|
> |[numpy](https://github.com/numpy/numpy) 	| 2.0 sec 	| 49 min 9 sec (1,475x slower)|
> | [scikit-learn](https://github.com/scikit-learn/scikit-learn) 	| 2.4 sec 	| 3 hr 5 min 33 sec (4,639x slower) |

What did you do differently to improve the speed? Which algorithms do you use?

**Reply by jsh9** [reply]
> Can you maybe provide a comparison of supported lint rules?

This is the currently supported linting rules: https://github.com/jsh9/pydoclint/blob/a1fda04f8589382092f9cb88fded560d9ecfc325/README.md#3-style-violation-codes

They are, on a high level, not much different from what darglint offers: https://github.com/terrencepreilly/darglint#error-codes

> because I have big plans on extending the linting capabilities to class and module docstrings

Darglint doesn't check class docstrings, which, in my opinion, is an omission.  Because Sphinx, when rendering a documentation HTML, is expecting to read info from the class docstring rather than the `__init__()` docstring.

As to module docstrings, they are already checked by [pydocstyle](https://www.pydocstyle.org/en/latest/error_codes.html) (error code: `D100 - Missing docstring in public module`).

In pydoclint, I don't plan to check any styles that pydocstyle already checks.

> What did you do differently to improve the speed? Which algorithms do you use?

I didn't do anything magical.

I'm using [numpydoc](https://github.com/jsh9/pydoclint/blob/a1fda04f8589382092f9cb88fded560d9ecfc325/setup.cfg#L20) to parse numpy-style docstrings, and [docstring_parser](https://github.com/jsh9/pydoclint/blob/a1fda04f8589382092f9cb88fded560d9ecfc325/setup.cfg#L21) to parse Google-style docstrings.  Sphinx-style docstrings can also be parsed by the latter parser.

And then I'm using Python's standard library `ast` to parse the Python code (the function signature including type hints, `raise`'s, `return`'s, `yield`'s).

All these tasks can be achieved very fast, even in pure Python, so I'm somewhat confused as to why darglint is so slow.

**Reply by jsh9** [reply]
> For that matter I would also be interested in the other features you have planned and the philosophy/vision of your tool.

As of now, I think I have implemented almost everything that I want to achieve in pydoclint, which includes:

- Ensure the the input args in the function signature match the "args" section in the docstring
- If there are `return`'s, `raise`'s, `yield`'s in the function body (or return type annotation), check the docstring for corresponding sections

And specifically, I don't plan to implement any checks that [pydocstyle](https://www.pydocstyle.org/en/latest/error_codes.html) already supports, because that would mean reinventing the wheels.

There may be other use cases that I haven't thought of yet.  If users raise such requests, I'll take a look and implement them if I can.

**Comment by akaihola** [comment]
Hi @jsh9, thanks for reaching out!

The story behind the Darglint2 fork is:
- I needed to unify and check docstrings in a project of mine.
- Out of the linters I found at the time, only Darglint did what I needed.
- Darglint was inconvenient to configure for my needs, but there was a fix in the open PR terrencepreilly/darglint#210
- @terrencepreilly had stopped maintenance of Darglint and archived the project.
- When asking Terrence about forking and taking over the PyPI entry, his preference was a renamed fork.
- Thus was born Darglint2, with the original intent to just publish it on PyPI with the "Parse ignore regexes from config" fix from @9dogs.
- @real-yfprojects stepped in and has been working hard on other improvements and clean-up on the project.

Regarding performance, Terrence was aware of the problem and had already experimented replacing the parser with an LR parser (see [darglint#186 comments](https://github.com/terrencepreilly/darglint/issues/186#issuecomment-1047316160)). There's also a [utility](/akaihola/darglint2/blob/master/integration_tests/performance.py) for measuring the tool's performance characteristics in the integration tests.

I think it definitely makes sense for Darglint2 documentation to point to pydoclint (it already [mentions pydocstyle](/akaihola/darglint2#scope)) and warn about the current performance issue.

If you guys decide to join forces and make pydoclint fill all the needs for which people use Darglint, I think the value of Darglint2 becomes to be a maintained fork of Darglint for those existing users who need the bugfixes but don't want to migrate to a different tool. In that case we should naturally point to maintained tools at the top of the Darglint2 README (It's a bit unfortunate that we can't achieve that for the README of the original Darglint [on PyPI](https://pypi.org/project/darglint/) nor [on GitHub](/terrencepreilly/darglint)).

**Reply by jsh9** [reply]
Thanks for the reply!

This PR (https://github.com/terrencepreilly/darglint/pull/210) reminded me that I should also parse options from a config file in pydoclint.

In my humble opinion, the performance of darglint/darglint2 can't be solved without a full rewrite or switching to other docstring parsers (which is what I did).

Another dealbreaker for me to use darglint is: in-line `# noqa` comments never work well.  On the other hand, pydoclint supports in-line `# noqa` natively (users can put it on the line of the function/method, or the line of the class).  I suspect fixing this issue in darglint2 will require big refactoring too.

So I suppose the efforts required for improving/fixing darglint may be much larger than improving/fixing pydoclint.  If you'd like to join forces in improving pydoclint, you are more than welcome!

**Reply by real-yfprojects** [reply]
> If you'd like to join forces in improving pydoclint, you are more than welcome!

While it feels sad leaving darglint2 behind I think pydoclint has more potential in its turrent state so putting my efford into pydoclint will be better.

I have just set up sphinx docs for darglint that are hosted on githeb pages and I am currently improving community standards and exteding the docs. See akaihola/darglint2#29. I could continue this task with pydoclint.

**Reply by real-yfprojects** [reply]
> I have just set up sphinx docs for darglint that are hosted on githeb pages and I am currently improving community standards and exteding the docs. See akaihola/darglint2#29. I could continue this task with pydoclint.

What do you think @jsh9?

**Reply by real-yfprojects** [reply]
The lack of response disheartens me and makes me reconsider my decision to help with *pydoclint*.

**Reply by akaihola** [reply]
@real-yfprojects don't take it personally, I know I have done the same sometimes when real life hits and I'm unable to work on my Open Source projects for many weeks in a row.

**Reply by real-yfprojects** [reply]
> don't take it personally, I know I have done the same sometimes when real life hits and I'm unable to work on my Open Source projects for many weeks in a row.

I wouldn't if he wasn't very active in other threads and even lively discussions in this repo while also rapidly implementing new features or bug fixes and pushing almost daily releases.

**Reply by jsh9** [reply]
Hi both!  I missed the notification from this thread somehow.  Too many emails in my inbox probably.  Sorry about that!

I see that @real-yfprojects asked me a question of whether I need a contributing guide or community standard for pydoclint.  Right now, I'm feeling neutral about this.  But if either of you would like to contribute to it, you are more than welcome :)

**Reply by real-yfprojects** [reply]
> I see that @real-yfprojects asked me a question of whether I need a contributing guide or community standard for pydoclint. 

Yes, I was thinking of extensive documentation for users, developers and maintainers in the form of a sphinx website hosted on Github Pages. I can do the writing and lead the way, however we will have to discuss the exact contents requiring you to answer some questions.

**Reply by jsh9** [reply]
Do you think the current README is sufficient?  Or a README.md + CONTRIBUTING.md + <something_else>.md in the repo?

I'm asking this because a dedicated site makes it a bit less convenient to maintain.

**Reply by real-yfprojects** [reply]
> Do you think the current README is sufficient? Or a README.md + CONTRIBUTING.md + <something_else>.md in the repo?
> 
> I'm asking this because a dedicated site makes it a bit less convenient to maintain.

Other projects have a fully fledged documentation since the amount of text makes the README very long and therefore complicates finding stuff. Meanwhile extensive documentation is important for users so that they can easily use the tool but especially for maintainers and contributors. It allows contributors to get started quickly and decreases the burden of writing patches. It allows maintainers to tell contributors know what to expect and the maintainer has less effort communicating with contributors. Documentation builds the basis for success and longevity of a project. Without writing things down others can't take over tasks since they lack your knowledge about the project/workflow/.. which you only kept in your head. It also makes it hard for others to become a maintainer when you have less time. (Exactly that problem hit darglint, we don't want to end like so many unmaintained FOSS projects)

**Reply by jsh9** [reply]
Yeah, I think it's beneficial to have a dedicate documentation site rather than just README.

Does the source code for the Github page site also live in the same repo?  If so, please feel free to create a PR.  I have never created Github pages before, so I wouldn't know how to do it.

Thanks!

**Reply by real-yfprojects** [reply]
> Yeah, I think it's beneficial to have a dedicate documentation site rather than just README.
> 
> Does the source code for the Github page site also live in the same repo?  

Yes, it does. In the past there had to be a separate branch with the page contents. However there is a new feature where this is no longer needed. The html render of the docs will be exclusively available through the repo specific Url while the source resides in the main branch. 

> If so, please feel free to create a PR.  I have never created Github pages before, so I wouldn't know how to do it.

:+1:

**Reply by wwuck** [reply]
Would it be easier to automatically publish on https://readthedocs.org/ like many other python projects do?

**Reply by jsh9** [reply]
If my understanding is correct, what @real-yfprojects  proposed would be auto publishable too.  I've used readthedocs.org before.  I think it's a good platform, but if Github offers something built-in, it's better to use Github's (for convenience, mainly).

**Reply by real-yfprojects** [reply]
> I think it's a good platform, but if Github offers something built-in, it's better to use Github's (for convenience, mainly)

+ it saves the nonprofit? RTD money.

**Reply by jsh9** [reply]
I created a site based on GitHub pages: https://jsh9.github.io/pydoclint/

Right now this site is quite static: I manually added all the contents there (some of them are from README.md, so now README.md is less busy).

**Reply by real-yfprojects** [reply]
Looks neat!

> Right now this site is quite static

What would you like to be dynamic? The commandline options could be generated from the source code and one could have a section detailing the software architecture that documents the important classes and functions.

**Reply by jsh9** [reply]
Actually there may not be much that needs to be made dynamic.  A detailed API documentation may be dynamic, but this is not a Python library, so users don't really need to read the API docs of the internal functions.
